In [57]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [ ]:
data = pd.read_csv('car_fuel_efficiency_2026.csv')


In [129]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


In [135]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


Question 1
There's one column with missing values. What is it?

'engine_displacement'
'horsepower'
'vehicle_weight'
'model_year'

answers: horsepower

In [136]:
data.horsepower.describe()

count    9123.000000
mean      254.446016
std        22.844613
min       171.000000
25%       239.000000
50%       254.000000
75%       270.000000
max       334.000000
Name: horsepower, dtype: float64

What's the median (50% percentile) for variable 'horsepower'?

- 204
- 254
- 304
- 354

answers : 254

In [130]:
categorical_columns = ['model_year']

categorical = {}

for c in categorical_columns:
    categorical[c] = list(df_train[c].value_counts().head().index)

In [233]:
#Shuffle the filtered dataset and create the split exactly as in the lecture 
df=data.copy()
n = len(df)
n_val = int(n * 0.2)
n_test = int(n * 0.2)
n_train = n - n_val - n_test

np.random.seed(42)
idx = np.arange(n)
np.random.shuffle(idx)

df_train = df.iloc[idx[:n_train]]
df_val = df.iloc[idx[n_train:n_train + n_val]]
df_test = df.iloc[idx[n_train + n_val:]]

In [139]:

#getting mean of the df train
mean_train = df_train.horsepower.mean()

In [140]:
mean_train

np.float64(254.46338337605272)

Question 3
We need to deal with missing values for the column from Q1.
We have two options: fill it with 0 or with the mean of this variable.
Try both options. For each, train a linear regression model without regularization using the code from the lessons.
For computing the mean, use the training only!
Use the validation dataset to evaluate the models and compare the RMSE of each option.
Round the RMSE scores to 3 decimal digits using round(score, 3). This keeps the imputation difference visible in this release.
Which option gives better RMSE?

In [234]:
#create two versions of df
df_train_v1 = df_train.copy()
df_val_v1 = df_val.copy()
df_test_v1 = df_test.copy()

df_train_v1 = df_train_v1.reset_index(drop=True)
df_val_v1 = df_val_v1.reset_index(drop=True)
df_test_v1 = df_test_v1.reset_index(drop=True)

In [235]:
df_train_v2 = df_train.copy()
df_val_v2 = df_val.copy()
df_test_v2 = df_test.copy()

df_train_v2 = df_train_v2.reset_index(drop=True)
df_val_v2 = df_val_v2.reset_index(drop=True)
df_test_v2 = df_test_v2.reset_index(drop=True)

In [170]:
#set up the linear regression functions
def train_linear_regression(X, y):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]



#set up RMSE Function
def rmse(y, y_pred):
    se = (y - y_pred) ** 2
    mse = se.mean()
    return np.sqrt(mse)

In [236]:
#set up the y value

y_train = np.log1p(df_train.fuel_efficiency_mpg.values)
y_val = np.log1p(df_val.fuel_efficiency_mpg.values)
y_test = np.log1p(df_test.fuel_efficiency_mpg.values)



In [237]:
del df_train_v1['fuel_efficiency_mpg']
del df_val_v1['fuel_efficiency_mpg']
del df_test_v1['fuel_efficiency_mpg']

del df_train_v2['fuel_efficiency_mpg']
del df_val_v2['fuel_efficiency_mpg']
del df_test_v2['fuel_efficiency_mpg']

In [173]:
#fill missing value with mean (version1) and with zero (version2)
df_train_v1['horsepower'] = df_train_v1['horsepower'].fillna(254.46338337605272)
df_val_v1['horsepower'] = df_val_v1['horsepower'].fillna(254.46338337605272)
df_test_v1['horsepower'] = df_test_v1['horsepower'].fillna(254.46338337605272)
df_train_v2['horsepower'] = df_train_v2['horsepower'].fillna(0)
df_val_v2['horsepower'] = df_val_v2['horsepower'].fillna(0)
df_test_v2['horsepower'] = df_test_v2['horsepower'].fillna(0)

In [238]:
def prepare_X(df):
    base = ['engine_displacement',
'horsepower',
'vehicle_weight']
    features = base.copy() 
    print(features)
    for name, values in categorical.items():
        for value in values:
            df['%s_%s' % (name, value)] = (df[name] == value).astype(int)
            features.append('%s_%s' % (name, value))
    print(features)
    df_num = df[features]
    df_num = df_num.fillna(0)
    X = df_num.values

    return X

In [239]:
#train the model for version 1
X_train_v1 = prepare_X(df_train_v1)

w0_v1, w_v1 = train_linear_regression(X_train_v1, y_train_v1)

y_pred_v1 = w0_v1 + X_train_v1.dot(w_v1)

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']


In [240]:
#check rmse for model version 1 
round(rmse(y_train_v1, y_pred_v1),3)

np.float64(0.083)

In [241]:
#train the model for version 2
X_train_v2 = prepare_X(df_train_v2)

w0_v2, w_v2 = train_linear_regression(X_train_v2, y_train_v2)

y_pred_v2 = w0_v2 + X_train_v2.dot(w_v2)

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']


In [242]:
#check rmse for model version 2 
round(rmse(y_train_v2, y_pred_v2),3)

np.float64(0.083)

Both model has the rsme, now we need to compare the rmse score using data for validation

In [243]:
#validate with model version 1
X_val_v1 = prepare_X(df_val_v1)

y_pred_v1 = w0_v1 + X_val_v1.dot(w_v1)
round(rmse(y_val_v1, y_pred_v1),3)

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']


np.float64(0.082)

In [245]:
#validate with model version 2
X_val_v2 = prepare_X(df_val_v2)

y_pred_v2 = w0_v2 + X_val_v2.dot(w_v2)
round(rmse(y_val_v2, y_pred_v2),3)

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']


np.float64(0.082)

Both model has the same RMSE score so both are equally good

Question 4
Now let's train a regularized linear regression.
For this question, fill the NAs with 0.
Try different values of r from this list: [0, 0.01, 0.1, 1, 5, 10, 100].
Use RMSE to evaluate the model on the validation dataset.
Round the RMSE scores to 4 decimal digits. This keeps the small but real regularization differences visible instead of turning several choices into a tie.
Which r gives the best RMSE?

In [246]:
#create function to train linreg with penalty
def train_linear_regression_reg(X, y, r=0.001):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX = XTX + r * np.eye(XTX.shape[0])

    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [247]:
for r in [0, 0.01, 0.1, 1, 5, 10, 100]:
    w0, w = train_linear_regression_reg(X_train_v2, y_train, r=r)

    X_val = prepare_X(df_val_v2)
    y_pred = w0 + X_val.dot(w)
    score = rmse(y_val_v2, y_pred)
    
    print(r, w0, round(score,3))

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
0 4.1045656368170285 0.082
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
0.01 4.102752084116563 0.082
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
0.1 4.086501984338479 0.082
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
1 3.9308151404400506 0.083
['engine_displacement', 'horsepower', 'vehicle_weight']
['eng

the answers for question 4 is 0, as the lowest number who give the lowest RMSE score is 0

Question 5
We used seed 42 for splitting the data. Let's find out how selecting the seed influences our score.
Try different seed values: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].
For each seed, do the train/validation/test split with 60%/20%/20% distribution.
Fill the missing values with 0 and train a model without regularization.
For each seed, evaluate the model on the validation dataset and collect the RMSE scores.
What's the standard deviation of all the scores? To compute the standard deviation, use np.std.
Round the result to 3 decimal digits (round(std, 3))

In [248]:
for rseed in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
    np.random.seed(rseed)
    idx = np.arange(n)
    np.random.shuffle(idx)

    print(idx)

[9394  898 2398 ... 3264 9845 2732]
[9953 3850 4962 ...  905 5192  235]
[7878 3224 1919 ... 6637 2575 7336]
[5876 6555 1448 ... 9859 1688 5994]
[1603 8713 4561 ...  709 8366 1146]
[7054  442 3954 ... 4079 2254 2915]
[ 739 9426 8200 ... 4714 8419 2761]
[1977 3880   52 ... 5699  537 9412]
[ 837  697 1197 ... 2033 9556 4547]
[3644 9184  520 ... 6782 4444 8574]


In [249]:
#train model for different seed values 
data = pd.read_csv('car_fuel_efficiency_2026.csv')


df=data.copy()

n = len(df)
n_val = int(n * 0.2)
n_test = int(n * 0.2)
n_train = n - n_val - n_test
df = df.fillna(0)
all_rmse_scores = []
for rseed in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
    np.random.seed(rseed)
    idx = np.arange(n)
    np.random.shuffle(idx)
    
    df_train = df.iloc[idx[:n_train]]
    df_val = df.iloc[idx[n_train:n_train + n_val]]
    df_test = df.iloc[idx[n_train + n_val:]]
   
    
    df_train = df_train.reset_index(drop=True)
    df_val = df_val.reset_index(drop=True)
    df_test = df_test.reset_index(drop=True)

    y_train = np.log1p(df_train.fuel_efficiency_mpg.values)
    y_val = np.log1p(df_val.fuel_efficiency_mpg.values)
    y_test = np.log1p(df_test.fuel_efficiency_mpg.values)

    del df_train['fuel_efficiency_mpg']
    del df_val['fuel_efficiency_mpg']
    del df_test['fuel_efficiency_mpg']
    
    #train the model 
    X_train = prepare_X(df_train)
    w0, w = train_linear_regression(X_train, y_train)
    #validate
    X_val = prepare_X(df_val)
    y_pred = w0 + X_val.dot(w)
    all_rmse_scores.append(rmse(y_val, y_pred))

['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'mo

(8000, 9)

In [250]:
all_rmse_scores

[np.float64(0.08430859898824729),
 np.float64(0.08194505264307504),
 np.float64(0.08106950681602414),
 np.float64(0.08371755984506776),
 np.float64(0.08161387407372808),
 np.float64(0.0847699314800465),
 np.float64(0.08509692147458199),
 np.float64(0.0819613365631581),
 np.float64(0.08274865918814178),
 np.float64(0.08233304687077896)]

In [254]:
#checking the std for all rmse scores
round(np.std(all_rmse_scores),3)

np.float64(0.001)

answers to question 5 : 0.01 (failed to find the correct answers)

Question 6
Split the dataset like previously, use seed 9.
Combine train and validation datasets.
Fill the missing values with 0 and train a model with r=0.001.
What's the RMSE on the test dataset?

In [252]:
df = pd.read_csv('car_fuel_efficiency_2026.csv')
np.random.seed(9)
idx = np.arange(n)
np.random.shuffle(idx)

#df_train = df.iloc[idx[:n_train]]
df_train_val = df.iloc[idx[:n_train + n_val]]
df_test = df.iloc[idx[n_train + n_val:]]


df_train = df_train_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

y_train = np.log1p(df_train.fuel_efficiency_mpg.values)
y_test = np.log1p(df_test.fuel_efficiency_mpg.values)

del df_train['fuel_efficiency_mpg']
del df_test['fuel_efficiency_mpg']
#train the model 
X_train = prepare_X(df_train)
w0, w = train_linear_regression_reg(X_train, y_train,r=0.001)
#validate the model
X_test =prepare_X(df_test)



['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']
['engine_displacement', 'horsepower', 'vehicle_weight']
['engine_displacement', 'horsepower', 'vehicle_weight', 'model_year_1990', 'model_year_1979', 'model_year_1984', 'model_year_2020', 'model_year_1995']


In [255]:
y_pred = w0 + X_test.dot(w)
rmse(y_test, y_pred)

np.float64(0.08385710005048222)

answers to question 6 : I got 0.083 (failed to find the correct answers)